_posts/2017-01-23-cpp-strings.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Modeling strings with libpmemobj C++ bindings
+author: RobDickinson
+layout: post
+---
+
+C++ developers using libpmemobj have more than one option for modeling strings,
+depending on the size of the strings and whether they are fixed or varying
+in length. In this post we'll review the representations that work,
+known variations to avoid, and finally present a persistent string
+class that implements these best practices.
+
+#### Avoid wrapping fixed-size arrays
+
+You might expect (like I did at first!) that `p<char[size]>` is a proper way
+to simply model a fixed-size string, but actually this is not correct.
+The `p<>` template does not implement the subscript `operator[]` and so
+`p<char[size]>` won't even compile.
+
+#### Avoid wrapping std::string
+
+Never use `p<std::string>` or `persistent_ptr<std::string>`. This code will
+compile, but these implementations will not be power-fail safe (at best)
+and might be unstable (at worst). This advice applies to using `p<>`
+or `persistent_ptr<>` wrappers around any complex types that perform
+their own memory management. Don't make this obvious mistake!
+
+#### Avoid arrays of persistent pointers to single chars
+
+Don't model a variable-length string using `persistent_ptr<char> X[]`. This
```

Comment:
Now that I think about it, this should be `persistent_ptr<p<char>>`. That is because `persistent_ptr` does not track the contents of the object. Sorry about not noticing this in the previous review. One more thing to note with this approach would be to `make_persistent` each and every array element which is evil. I was also thinking about `persistent_ptr<p<char>[]>` which could actually work. It would be inefficient, you would have to assign each character separately through `operator[]` (`get_rw()` could be a workaround for this) and `operator=`, but it should actually be powerfail safe.